src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,66 +1,120 @@
-			}],
-		})
-		return true
+	ingestPayloadStream = Writable({
+		objectMode: true,
+		write: async (payload, enc, done) => {
+			if (!(payload instanceof Payload)) {
+				done(new Error('ingestPayloadStream received non-Payload data'))
+			} else {
+				done(this.producer.send({
+					topic: dataTypes.STREAM.CONTAINER,
+					messages: [{
+						value: Payload.serialize(payload),
+					}],
+				}))
+			}
+		},
+	})
+
+	/**
+	 * Updates ingestion state based on the latest demuxed packet data.
+	 *
+	 * This method is called by our mpegts demuxer, and allows the ingestion engine to track
+	 * the most recent demuxed packet.
+	 *
+	 * @param  {Packet} packet The latest TSDemuxer Packet object.
+	 */
+	processDemuxedPacket = (packet) => {
+		this.mostRecentDemuxedPacket = packet
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Returns the most recent cohearant stream packet processed by the ingestion engine.
+	 * This packet is lower level than the mpegts container, and represents an audio or video
+	 * packet demuxed from the mpegts stream.
+	 *
+	 * @return {Packet} The most recent Packet object extracted by TSDemuxer.
+	 */
+	getMostRecentDemuxedPacket = () => this.mostRecentDemuxedPacket
+
+	/**
+	 * Process a new chunk of data from an mpegts stream. The chunks passed to this
+	 * function should be presented sequentially, and combine to form a cohearant mpegts
+	 * data stream, but they can be of arbitray size.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
+	 * This data is used to update local state and ultimately pipe to kafka.
+	 *
+	 * @param  {Buffer} mpegtsData The latest sequential chunk of mpegts data.
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
-		})
+	processMpegtsStream = Transform({
+		objectMode: true,
+		write: (mpegtsData, enc, done) => {
+			this.mpegtsDemuxer.process(mpegtsData)
+			const demuxedPacket = this.getMostRecentDemuxedPacket() || { pts: 0 }
+			const payload = new Payload({
+				data: mpegtsData,
+				ptsTime: tsToSeconds(demuxedPacket.pts),
+			})
+			done(null, payload)
+		},
+	})
 
 	/**
 	 * Start the ingestion engine.
 	 *
-	 * @return {null}
+	 * This creates an ffmpeg process configured to convert input into a predetermined
+	 * format as specified by getFfmpegSettings, and announce new data to the ingestion
+	 * engine.
+	 *
+	 * It then spawns the input stream as specified by the implementing class, and creates
+	 * a pipeline that passes input->ffmpeg->packaging->ingestion
+	 *
+	 * @return {Stream} the ingestion pipeline that just got started
 	 */
 	start = () => {
-		const ffmpegSettings = this.getFfmpegSettings()
-		const ffmpegProcess = spawn('ffmpeg', ffmpegSettings)
+		const ffmpeg = spawn('ffmpeg', this.getFfmpegSettings())
```

Comment:
Great point.

It *shouldn't* cause testing hangs in `yarn test` style testing, since we mock spawn, spawn never actually gets called during the test suite.

Is it hanging during the tests / timing out?  Or something else.

Reply:
Yeah, sorry. Not in yarn test. I did my own ad hoc integration testing so I didn’t send you something broken.

Reply:
I admit I didn't replicate this first, and we really should make integration tests, but I *did* add a `stop` method that gets called at the end of the pipeline so hopefully it's fixed?